File: frontend/src/llm/chr_file.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum LanguageSelector {
    #[serde(rename = "en")]
    English,
    #[serde(rename = "zh_CN")]
    Chinese,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusBarConfig {
    pub format: String,
    pub rule: Option<String>,
    pub sections: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SimulatorConfig {
    pub tasks: Option<String>,
    pub commands: Option<String>,
    pub world: Option<String>,
    pub characters: Option<String>,
    pub database: Option<String>,
    pub behaviors: Option<String>,
    pub prohibitions: Option<String>,
    pub sections: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemorySummarizerConfig {
    pub rules: Option<String>,
    pub sections: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SimulatorCHR {
    pub universe_name: String,
    pub literal_work_name: String,
    pub prologue: Option<String>,

    pub language: LanguageSelector,

    pub status_bar: StatusBarConfig,
    pub simulator: Option<SimulatorConfig>,
    pub memory: Option<MemorySummarizerConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdditionalCHR {
    pub status_bar: Option<StatusBarConfig>,
    pub simulator: Option<SimulatorConfig>,
    pub memory: Option<MemorySummarizerConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlayerCHR {
    pub player_name: String,
    pub settings: Option<String>,
}

File: frontend/src/llm/chr_file.ts
import type { LanguageSelector } from './lang_config'

export interface CHRBase {}

export interface SimulatorCHR extends CHRBase {
    universeName: string
    literalWorkName: string
    prologue?: string

    language: LanguageSelector

    statusBar: StatusBarConfig
    simulator?: SimulatorConfig
    memory?: MemorySummarizerConfig
}

export interface AdditionalCHR extends CHRBase {
    statusBar?: StatusBarConfig
    simulator?: SimulatorConfig
    memory?: MemorySummarizerConfig
}

export interface PlayerCHR extends CHRBase {
    playerName: string
    settings?: string
}

export type CHR = SimulatorCHR | AdditionalCHR | PlayerCHR

export interface CHRFile<C extends CHRBase> {
    fileName: string
    absolutePath: string
    chr: C
}

export interface StatusBarConfig {
    format: string
    rule?: string
    sections?: string
}

export interface SimulatorConfig {
    tasks?: string
    commands?: string
    world?: string
    characters?: string
    database?: string
    behaviors?: string
    prohibitions?: string
    sections?: string
}

export interface MemorySummarizerConfig {
    rules?: string
    sections?: string
}
